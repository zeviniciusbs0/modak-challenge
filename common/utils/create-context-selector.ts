import { type Context, useContextSelector } from "use-context-selector";

export const createContextSelector =
	<T>(context: Context<T>) =>
	<Selected>(selector: (state: T) => Selected) => {
		return useContextSelector(context, selector);
	};
