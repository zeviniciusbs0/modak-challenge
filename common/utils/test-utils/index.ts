// Re-export everything from React Native Testing Library
export * from "@testing-library/react-native";

// Re-export common Jest functions
export { jest } from "@jest/globals";

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
	renderHook as originalRenderHook,
	render,
} from "@testing-library/react-native";
import type {
	RenderHookOptions,
	RenderOptions,
} from "@testing-library/react-native";
// Custom render function with providers
import React from "react";
import { TamaguiProvider } from "tamagui";
import { tamaguiConfig } from "../../../tamagui.config";

// Create QueryClient with test-friendly defaults
const createTestQueryClient = () =>
	new QueryClient({
		defaultOptions: {
			queries: {
				retry: false,
				staleTime: Number.POSITIVE_INFINITY,
				gcTime: 0, // Immediate garbage collection
			},
			mutations: {
				retry: false,
			},
		},
	});

// Create a custom render function that includes providers
const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	const queryClient = createTestQueryClient();

	return React.createElement(
		TamaguiProvider,
		{ config: tamaguiConfig, defaultTheme: "light" },
		React.createElement(QueryClientProvider, { client: queryClient }, children),
	);
};

const customRender = (
	ui: React.ReactElement,
	options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: Wrapper, ...options });

const customRenderHook = <Result, Props>(
	hook: (initialProps: Props) => Result,
	options?: Omit<RenderHookOptions<Props>, "wrapper">,
) => originalRenderHook(hook, { wrapper: Wrapper, ...options });

// Re-export everything, overriding render and adding renderHook
export { customRender as render, customRenderHook as renderHook };

// Test utilities
export const createMockNavigation = () => ({
	navigate: jest.fn(),
	goBack: jest.fn(),
	dispatch: jest.fn(),
	reset: jest.fn(),
	canGoBack: jest.fn(() => true),
	isFocused: jest.fn(() => true),
});

export const createMockRoute = (params = {}) => ({
	key: "test-route",
	name: "TestScreen",
	params,
});

// Helper to wait for async operations
export const waitForNextTick = () =>
	new Promise((resolve) => setTimeout(resolve, 0));

// Mock data generators
export const createMockProduct = (overrides = {}) => ({
	id: 1,
	title: "Test Product",
	description: "Test Description",
	price: 99.99,
	category: "test-category",
	thumbnail: "https://test.com/image.jpg",
	images: ["https://test.com/image.jpg"],
	rating: 4.5,
	stock: 10,
	tags: ["test"],
	brand: "Test Brand",
	sku: "TEST-SKU",
	weight: 1,
	dimensions: { width: 10, height: 10, depth: 10 },
	warrantyInformation: "Test warranty",
	shippingInformation: "Test shipping",
	availabilityStatus: "In Stock",
	reviews: [],
	returnPolicy: "Test return",
	minimumOrderQuantity: 1,
	meta: {
		createdAt: "2024-01-01",
		updatedAt: "2024-01-01",
		barcode: "test-barcode",
		qrCode: "test-qr",
	},
	discountPercentage: 0,
	...overrides,
});

export const createMockCategory = (overrides = {}) => ({
	slug: "test-category",
	name: "Test Category",
	url: "https://dummyjson.com/products/category/test-category",
	...overrides,
});
