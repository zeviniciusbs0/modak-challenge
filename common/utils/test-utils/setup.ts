jest.mock("@/common/config/notifications", () => ({
	useConfigureNotificationsHandler: jest.fn(() => ({
		expoPushToken: "ExponentPushToken[test]",
		notification: null,
	})),
	sendLocalNotification: jest.fn(),
}));

jest.mock("axios", () => ({
	create: jest.fn(() => ({
		get: jest.fn(),
		post: jest.fn(),
		put: jest.fn(),
		delete: jest.fn(),
	})),
	get: jest.fn(),
	post: jest.fn(),
	put: jest.fn(),
	delete: jest.fn(),
}));

global.console = {
	...console,
	warn: jest.fn(),
	error: jest.fn(),
};

afterEach(() => {
	jest.runOnlyPendingTimers();
	jest.useRealTimers();
});

afterAll(() => {
	return new Promise((resolve) => setTimeout(resolve, 100));
});
