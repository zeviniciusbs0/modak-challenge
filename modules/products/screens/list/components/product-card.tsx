import { CurrencyHandler } from "@/common/utils/currency";
import type { Product } from "@/modules/products/types/product";
import { router } from "expo-router";
import { Card, H3, Image, Paragraph } from "tamagui";

type ProductCardProps = {
	product: Product;
};

export const ProductCard = ({ product }: ProductCardProps) => {
	const handlePress = () => {
		router.push(`/products/${product.id}`);
	};

	return (
		<Card elevate onPress={handlePress} size="$4" bordered my="$2">
			<Card.Header>
				<Image
					source={{ uri: product.thumbnail }}
					flex={1}
					width="100%"
					height={200}
				/>
			</Card.Header>
			<Card.Footer flexDirection="column" padded>
				<Paragraph color="$accent2" numberOfLines={2}>
					{product.title}
				</Paragraph>
				<H3 color="$red10">{CurrencyHandler.format(product.price)}</H3>
			</Card.Footer>
		</Card>
	);
};
