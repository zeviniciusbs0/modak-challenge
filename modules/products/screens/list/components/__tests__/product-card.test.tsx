import { createMockProduct, render } from "@/common/utils/test-utils";
import { ProductCard } from "../product-card";

describe("ProductCard", () => {
	it("should render the product card correctly", () => {
		const { toJSON } = render(<ProductCard product={createMockProduct()} />);

		expect(toJSON()).toMatchSnapshot();
	});

	it("should calculate review count correctly", () => {
		const productWithManyReviews = createMockProduct({
			rating: 3.8,
			reviews: new Array(15).fill({ rating: 4, comment: "Good" }),
		});

		const { getByText } = render(
			<ProductCard product={productWithManyReviews} />,
		);

		// Testa a lógica de cálculo: rating + (reviews.length)
		expect(getByText("3.8 (15)")).toBeTruthy();
	});

	it("should handle empty reviews array", () => {
		const productWithoutReviews = createMockProduct({
			rating: 0,
			reviews: [],
		});

		const { getByText } = render(
			<ProductCard product={productWithoutReviews} />,
		);

		// Edge case importante: produtos sem reviews
		expect(getByText("0 (0)")).toBeTruthy();
	});
});
