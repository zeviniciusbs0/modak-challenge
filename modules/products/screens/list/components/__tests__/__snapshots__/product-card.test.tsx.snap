// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should render the product card correctly 1`] = `
<View
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "hsla(0, 0%, 95%, 1)",
      "borderBottomColor": "hsla(0, 0%, 88%, 1)",
      "borderBottomLeftRadius": 9,
      "borderBottomRightRadius": 9,
      "borderBottomWidth": 1,
      "borderLeftColor": "hsla(0, 0%, 88%, 1)",
      "borderLeftWidth": 1,
      "borderRightColor": "hsla(0, 0%, 88%, 1)",
      "borderRightWidth": 1,
      "borderStyle": "solid",
      "borderTopColor": "hsla(0, 0%, 88%, 1)",
      "borderTopLeftRadius": 9,
      "borderTopRightRadius": 9,
      "borderTopWidth": 1,
      "flexDirection": "column",
      "marginBottom": 7,
      "marginTop": 7,
      "position": "relative",
      "shadowColor": "rgb(0,0,0)",
      "shadowOffset": {
        "height": 12,
        "width": 0,
      },
      "shadowOpacity": 0.0392156862745098,
      "shadowRadius": 24,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "transparent",
        "flexDirection": "column",
        "marginBottom": "auto",
        "paddingBottom": 18,
        "paddingLeft": 18,
        "paddingRight": 18,
        "paddingTop": 18,
        "zIndex": 10,
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://test.com/image.jpg",
        }
      }
      style={
        {
          "flex": 1,
          "height": 200,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      {
        "backgroundColor": "transparent",
        "flexDirection": "column",
        "marginBottom": 0,
        "marginTop": "auto",
        "paddingBottom": 18,
        "paddingLeft": 18,
        "paddingRight": 18,
        "paddingTop": 18,
        "zIndex": 5,
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        {
          "color": "hsla(0, 0%, 8%, 1)",
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "300",
          "letterSpacing": 0,
          "lineHeight": 24,
          "userSelect": "auto",
        }
      }
      suppressHighlighting={true}
    >
      Test Product
    </Text>
    <View
      style={[]}
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={20}
        bbWidth={20}
        fill="hsl(50, 100%, 48.5%)"
        focusable={false}
        height={20}
        meetOrSlice={0}
        minX={0}
        minY={0}
        stroke="hsl(50, 100%, 48.5%)"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "color": "hsl(50, 100%, 48.5%)",
            },
            {
              "flex": 0,
              "height": 20,
              "width": 20,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={20}
      >
        <RNSVGGroup
          color="hsl(50, 100%, 48.5%)"
          fill={
            {
              "payload": 4294430208,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4294430208,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="2"
        >
          <RNSVGPath
            d="M12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2z"
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
              ]
            }
            stroke={
              {
                "payload": 4294430208,
                "type": 0,
              }
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <Text
        style={[]}
      >
        4.5
         (
        0
        )
      </Text>
    </View>
    <Text
      style={[]}
    >
      $99.99
    </Text>
  </View>
</View>
`;
